OpenRA.Mods.Common/AI/BaseBuilder.cs

```diff
@@ -153,33 +177,76 @@ ActorInfo ChooseBuildingToBuild(ProductionQueue queue)
 			if (!ai.HasAdequateProc() || !ai.HasMinimumProc())
 			{
 				var refinery = GetProducibleBuilding("Refinery", buildableThings);
-				if (refinery != null)
+				if (refinery != null && SufficientPowerForActor(refinery))
 				{
 					HackyAI.BotDebug("AI: {0} decided to build {1}: Priority override (refinery)", queue.Actor.Owner, refinery.Name);
 					return refinery;
 				}
+
+				if (power != null && refinery != null && !SufficientPowerForActor(refinery))
+				{
+					HackyAI.BotDebug("{0} decided to build {1}: Priority override (would be low power)", queue.Actor.Owner, power.Name);
+					return power;
+				}
 			}
 
 			// Make sure that we can can spend as fast as we are earning
 			if (ai.Info.NewProductionCashThreshold > 0 && playerResources.Resources > ai.Info.NewProductionCashThreshold)
 			{
 				var production = GetProducibleBuilding("Production", buildableThings);
-				if (production != null)
+				if (production != null && SufficientPowerForActor(production))
 				{
 					HackyAI.BotDebug("AI: {0} decided to build {1}: Priority override (production)", queue.Actor.Owner, production.Name);
 					return production;
 				}
+
+				if (power != null && production != null && !SufficientPowerForActor(production))
+				{
+					HackyAI.BotDebug("{0} decided to build {1}: Priority override (would be low power)", queue.Actor.Owner, power.Name);
+					return power;
+				}
+			}
+
+			// Make sure that we can can spend as fast as we are earning
+			// Only consider building this if the last ChooseBuildLocation check did not fail
+			if (waterAvailable && ai.Info.NewProductionCashThreshold > 0 && playerResources.Resources > ai.Info.NewProductionCashThreshold)
+			{
+				var navalproduction = GetProducibleBuilding("NavalProduction", buildableThings);
+				if (navalproduction != null && SufficientPowerForActor(navalproduction))
+				{
+					var navalBuilding = navalproduction.ToString();
+					var buildLoc = ai.ChooseBuildLocation(navalBuilding, true, BuildingType.Building);
+					if (buildLoc != null)
+					{
+						HackyAI.BotDebug("AI: {0} decided to build {1}: Priority override (navalproduction)", queue.Actor.Owner, navalproduction.Name);
+						return navalproduction;
+					}
+					else
+						waterAvailable = false;
```

Comment:
If you haven't expanded to the full range of your base radius then it may be possible to place a naval yard later even if you can't place one right now.